cranelift/docs/isle.md

```diff
@@ -0,0 +1,1306 @@
+ISLE. Any term `A` that has one or more `(rule (A ...) RHS)` rules in
+the ISLE source implicitly has an internal constructor, and this
+constructor can be invoked from the right-hand side of other rules.
+
+#### Extractors
+
+Extractor behavior on a term allows it to be used in a pattern in the
+left-hand side of a rule. If one considers a constructor to be a
+function that goes from argument values to the complete term, then an
+extractor is a function that takes a complete term and possibly
+matches on it (it is fallible). If it does match, it provides the
+arguments *as results*.
+
+One can see extractors as "programmable match operators". They are a
+generalization of enum-variant deconstruction. Where a traditional
+term-rewriting system operates on a term data-structure that exists in
+memory, and can discover that a pattern `(A x y)` matches a term `A`
+at a particular point in the input, an extractor-based system instead
+sees `A` as an *arbitrary programmable operator* that is invoked
+wherever a pattern-match is attempted, and can return "success" with
+the resulting "fields" as if it were actually an enum variant. For
+more on this topic, see the motivation and description in [RFC 15
+under "programmable matching on virtual
+nodes"](https://github.com/bytecodealliance/rfcs/blob/main/accepted/cranelift-isel-isle-peepmatic.md#extractors-programmable-matching-on-virtual-nodes).
+
+To provide a concrete example, if we have the term declarations
+
+```lisp
+    (decl A (u32 u32) T)
+    (decl B (T) U)
+```
+
+then if we write a rule like
+
+```lisp
+    (rule (B (A x y))
+          (U.Variant1 x y))
+```
+
+then we have used `A` as an *extractor*. When `B` is invoked as a
+constructor with some `T`, this rule uses `A` as an extractor and
+attempts (via whatever programmable matching behavior) to use `A` to
+turn the `T` into two `u32`s, binding `x` and `y`. `A` can succeed or
+fail, just as any other part of a pattern-match can.
+
+Just as for constructors, there are *internal* and *external*
+extractors. Most of the interesting programmable behavior occurs in
+external extractors, which are defined in Rust; we will discuss this
+further in a section below. Internal extractors, in contrast, behave
+like macros, and can be defined for convenience: for example, we can
+write
+
+```lisp
+    (decl A (u32 u32) T)
+    (extractor (A pat1 pat2)
+               (and
+                 (extractArg1 pat1)
+                 (extractArg2 pat2)))
+```
+
+which will, for example, expand a pattern `(A (subterm ...) _)` into
+`(and (extractArg1 (subterm ...)) (extractArg2 _))`.
+
+#### Summary: Terms, Constructors, and Extractors
+
+We start with a `term`, which is just a schema for data:
+
+```lisp
+    (decl Term (A B C u32 u32) T)
+```
+
+A term can have:
+
+1. A single internal extractor body, via a toplevel `(extractor ...)`
+   form, OR
+   
+2. A single external extractor binding, via an `(extern extractor
```

Comment:
nit: the `(extern ...)` syntax was not introduced above but is here in the "summary", I would leave the syntax for the next section

Reply:
Updated, thanks.